File: duckdb_profiler/duckdb_planbench/prune.py
from typing import Dict, Any


def prune_projects_one_child(node: Dict[str, Any]) -> Dict[str, Any]:
    kids = [prune_projects_one_child(ch) for ch in node.get("children", [])]
    node = dict(node)
    node["children"] = kids
    if node.get("op") == "Project" and len(kids) == 1:
        child = dict(kids[0])
        if node.get("_slot_lineage") is not None:
            child["_slot_lineage"] = node["_slot_lineage"]
        return child
    return node


def strip_internal(node: Dict[str, Any]) -> Dict[str, Any]:
    out = {k: v for k, v in node.items() if k != "_slot_lineage"}
    out["children"] = [strip_internal(ch) for ch in node.get("children", [])]
    return out


def prune_cte(node: Dict[str, Any]) -> Dict[str, Any]:
    kids = [prune_cte(ch) for ch in node.get("children", [])]

    if node.get("op") == "CTE":
        if len(kids) == 1:
            return kids[0]
        out = dict(node)
        out["children"] = kids
        return out

    out = dict(node)
    out["children"] = kids
    return out

File: datafusion_profiler/src/prune.rs

use crate::schema::Node;

/// Remove `Project` nodes with exactly one child
pub fn prune_projects_one_child(node: Node) -> Node {
    let kids = node
        .children
        .into_iter()
        .map(prune_projects_one_child)
        .collect::<Vec<_>>();
    let mut node = Node { children: kids, ..node };
    if node.op == "Project" && node.children.len() == 1 {
        let mut child = node.children.remove(0);
        if let Some(sl) = node._slot_lineage.clone() {
            child._slot_lineage = Some(sl);
        }
        return child;
    }
    node
}

/// Remove internal `_slot_lineage` and infra_ms_accumulated from the final tree for cleanliness.
pub fn strip_internal(mut node: Node) -> Node {
    // Recurse
    let kids = node.children.into_iter().map(strip_internal).collect::<Vec<_>>();

    // Drop internal slot lineage
    node._slot_lineage = None;
    node.children = kids;

    // Strip infra accumulation field
    node.metrics.infra_ms_accumulated = None;

    node
}
